### PR TITLE
[enhancement](optimizer) Filter partition by hasData in pruner

### PR DESCRIPTION
**Pull request by Kikyou1997**
To reduce unnecessary rpc invocation

## Proposed changes

Issue Number: close #xxx

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by Kikyou1997** [comment]
run buildall

**Comment by wm1581066** [comment]
XX

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PruneOlapScanPartition.java [hunk 1]
```diff
@@ -52,7 +52,12 @@ public Rule build() {
             OlapTable table = scan.getTable();
             Set<String> partitionColumnNameSet = Utils.execWithReturnVal(table::getPartitionColumnNames);
             if (partitionColumnNameSet.isEmpty()) {
-                return filter;
+                return new LogicalFilter<>(filter.getConjuncts(),
+                        scan.withSelectedPartitionIds(scan.getSelectedPartitionIds().stream()
+                                .filter(
+                                        partitionId
+                                                -> table.getPartition(partitionId).hasData())
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PruneOlapScanPartition.java:59]: merge to one row

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PruneOlapScanPartition.java [hunk 2]
```diff
@@ -67,7 +72,8 @@ public Rule build() {
 
             List<Long> prunedPartitions = new ArrayList<>(PartitionPruner.prune(
                     partitionSlots, filter.getPredicate(), partitionInfo, ctx.cascadesContext,
-                    PartitionTableType.OLAP));
+                    PartitionTableType.OLAP)).stream()
+                    .filter(partitionId -> table.getPartition(partitionId).hasData()).collect(Collectors.toList());
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PruneOlapScanPartition.java:76]:
```suggestion
                    PartitionTableType.OLAP)).stream()
                    .filter(partitionId -> table.getPartition(partitionId).hasData())
                    .collect(Collectors.toList());
```

**Comment by Kikyou1997** [comment]
run buildall

**Comment by Kikyou1997** [comment]
run buildall

**Review by xzj7019 — APPROVED** [review]

**Review by morrySnow — APPROVED** [review]

**Pull request merged**